File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/AdditiveResponse.java
package openfoodfacts.github.scrachx.openfood.models;

import java.util.ArrayList;
import java.util.Map;

/**
 * Created by Lobster on 04.03.18.
 */

public class AdditiveResponse {

    private String tag;

    private Map<String, String> names;

    private String wikiDataCode;
    private Boolean isWikiDataIdPresent;
    private String overexposureRisk;
    private String exposureMeanGreaterThanAdi;
    private String exposureMeanGreaterThanNoael;
    private String exposure95ThGreaterThanAdi;
    private String exposure95ThGreaterThanNoael;

    public AdditiveResponse( String tag, Map<String, String> names, String overexposureRisk, String wikiDataCode )
    {
        this.tag = tag;
        this.names = names;
        this.wikiDataCode = wikiDataCode;
        this.overexposureRisk = overexposureRisk;
        this.isWikiDataIdPresent = true;
    }

    public AdditiveResponse( String tag, Map<String, String> names, String overexposureRisk )
    {
        this.tag = tag;
        this.names = names;
        this.overexposureRisk = overexposureRisk;
        this.isWikiDataIdPresent = false;
    }

    public void setExposureEvalMap( String exposure95ThGreaterThanAdi, String exposure95ThGreaterThanNoael, String exposureMeanGreaterThanAdi, String exposureMeanGreaterThanNoael )
    {
        this.exposure95ThGreaterThanAdi = exposure95ThGreaterThanAdi;
        this.exposure95ThGreaterThanNoael = exposure95ThGreaterThanNoael;
        this.exposureMeanGreaterThanAdi = exposureMeanGreaterThanAdi;
        this.exposureMeanGreaterThanNoael = exposureMeanGreaterThanNoael;
    }

    public Additive map() {
        Additive additive;

        if (isWikiDataIdPresent) {
            additive = new Additive(tag, new ArrayList<>(), overexposureRisk, wikiDataCode);
            for (Map.Entry<String, String> name : names.entrySet()) {
                AdditiveName additiveName = new AdditiveName(additive.getTag(), name.getKey(), name.getValue(), overexposureRisk, wikiDataCode);
                additiveName.setExposureEvalMap( exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael );
                additive.getNames().add(additiveName);
            }
        } else {
            additive = new Additive(tag, new ArrayList<>(), overexposureRisk );
            for (Map.Entry<String, String> name : names.entrySet()) {
                AdditiveName additiveName = new AdditiveName( additive.getTag(), name.getKey(), name.getValue(), overexposureRisk );
                additiveName.setExposureEvalMap( exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael );
                additive.getNames().add( additiveName );
            }
        }

        additive.setExposureEvalMap( exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael );

        return additive;
    }

}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/additive/AdditiveResponse.kt
package openfoodfacts.github.scrachx.openfood.models.entities.additive

import openfoodfacts.github.scrachx.openfood.models.entities.EntityResponse

/**
 * Created by Lobster on 04.03.18.
 */
class AdditiveResponse(
        private val tag: String,
        private val names: Map<String, String>,
        private val overexposureRisk: String?,
        private var wikiDataCode: String? = null
) : EntityResponse<Additive> {
    private var exposureMeanGreaterThanAdi: String? = null
    private var exposureMeanGreaterThanNoael: String? = null
    private var exposure95ThGreaterThanAdi: String? = null
    private var exposure95ThGreaterThanNoael: String? = null


    fun setExposureEvalMap(
            exposure95ThGreaterThanAdi: String?,
            exposure95ThGreaterThanNoael: String?,
            exposureMeanGreaterThanAdi: String?,
            exposureMeanGreaterThanNoael: String?
    ) {
        this.exposure95ThGreaterThanAdi = exposure95ThGreaterThanAdi
        this.exposure95ThGreaterThanNoael = exposure95ThGreaterThanNoael
        this.exposureMeanGreaterThanAdi = exposureMeanGreaterThanAdi
        this.exposureMeanGreaterThanNoael = exposureMeanGreaterThanNoael
    }

    override fun map(): Additive {
        val additive: Additive
        if (wikiDataCode != null) {
            additive = Additive(tag, arrayListOf(), overexposureRisk, wikiDataCode)
            names.forEach { (key, value) ->
                val additiveName = AdditiveName(additive.tag, key, value, overexposureRisk, wikiDataCode)
                additiveName.setExposureEvalMap(exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael)
                additive.names.add(additiveName)
            }
        } else {
            additive = Additive(tag, arrayListOf(), overexposureRisk)
            names.forEach { (key, value) ->
                val additiveName = AdditiveName(additive.tag, key, value, overexposureRisk)
                additiveName.setExposureEvalMap(exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael)
                additive.names.add(additiveName)
            }
        }
        additive.setExposureEvalMap(exposure95ThGreaterThanAdi, exposure95ThGreaterThanNoael, exposureMeanGreaterThanAdi, exposureMeanGreaterThanNoael)
        return additive
    }
}